src/vmm/src/snapshot/mod.rs

```diff
@@ -122,37 +122,22 @@ impl Snapshot {
             .map_err(|err| SnapshotError::Serde(err.to_string()))
     }
 
-    /// Helper function to serialize an object to a writer
-    pub fn serialize<T, O>(writer: &mut T, data: &O) -> Result<(), SnapshotError>
+    pub fn load_unchecked<R: Read>(reader: &mut R) -> Result<Self, SnapshotError>
     where
-        T: Write,
-        O: Serialize + Debug,
-    {
-        bincode::serialize_into(writer, data).map_err(|err| SnapshotError::Serde(err.to_string()))
-    }
-
-    /// Attempts to load an existing snapshot without performing CRC or version validation.
-    ///
-    /// This will check that the snapshot magic value is correct.
-    fn unchecked_load<T, O>(reader: &mut T) -> Result<(O, Version), SnapshotError>
-    where
-        T: Read + Debug,
-        O: DeserializeOwned + Debug,
+        Data: DeserializeOwned + Debug,
     {
         let hdr: SnapshotHdr = Self::deserialize(reader)?;
         if hdr.magic != SNAPSHOT_MAGIC_ID {
             return Err(SnapshotError::InvalidMagic(hdr.magic));
         }
 
-        let data: O = Self::deserialize(reader)?;
-        Ok((data, hdr.version))
+        let data: Data = Self::deserialize(reader)?;
+        Ok(Self { header: hdr, data })
     }
 
-    /// Load a snapshot from a reader and validate its CRC
-    pub fn load<T, O>(reader: &mut T, snapshot_len: usize) -> Result<(O, Version), SnapshotError>
+    pub fn load<R: Read>(reader: &mut R, snapshot_len: usize) -> Result<Self, SnapshotError>
```

Comment:
This function's implementation we can change to only do a single pass over the snapshot bytes, as well as not taking a `snapshot_len: usize` parameter, I think:

1. Construct `crc_reader` from the given reader, as is done now
2. Deserialize the `Snapshot<Data>` object from this reader. Since the `struct Snapshot` does not contain the `crc`, and `allow_trailing_bytes` is set for `bincode`, this will simply leave the crc bytes unread.
3. at the end, store the crc from the `CrcReader` in some local variable.
4. Read the remaining 4 bytes in the input stream, and check that they match the computed CRC.